contracts/bonding/BondingManager.sol

```diff
@@ -248,50 +253,44 @@ contract BondingManager is IBondingManager, Manager {
 
     /*
      * @dev Unbond delegator's current stake. Delegator enters unbonding state
+     * @param _amount Amount of tokens to unbond
      */
-    function unbond() external whenSystemNotPaused returns (bool) {
-        // Current round must be initialized
-        require(roundsManager().currentRoundInitialized());
-
-        Delegator storage del = delegators[msg.sender];
-
+    function unbond(uint256 _amount)
+        external
+        whenSystemNotPaused
+        currentRoundInitialized
+        updateDelegatorStake
+        returns (bool)
+    {
         // Sender must be in bonded state
         require(delegatorStatus(msg.sender) == DelegatorStatus.Bonded);
 
-        // Update delegator stake
-        updateDelegatorStake(msg.sender);
+        Delegator storage del = delegators[msg.sender];
 
         // Transition to unbonding phase
         del.withdrawRound = roundsManager().currentRound().add(unbondingPeriod);
 
-        // Decrease transcoder total stake
-        transcoderPools.decreaseTranscoderStake(del.transcoderAddress, del.bondedAmount);
+        if (transcoderStatus(del.delegationTarget) == TranscoderStatus.Registered) {
+            // Decrease transcoder total stake
+            transcoderPools.decreaseTranscoderStake(del.delegationTarget, del.bondedAmount);
+        }
 
         // Delegator no longer bonded to anyone
-        del.transcoderAddress = address(0);
+        del.delegationTarget = address(0);
 
         return true;
     }
 
     /**
      * @dev Withdraws withdrawable funds back to the caller after unbonding period.
      */
-    function withdraw() external whenSystemNotPaused returns (bool) {
-        // Current round must be initialized
-        require(roundsManager().currentRoundInitialized());
-
-        if (transcoderStatus(msg.sender) == TranscoderStatus.Unbonded) {
-            token.transfer(msg.sender, transcoders[msg.sender].bondedAmount);
+    function withdraw() external whenSystemNotPaused currentRoundInitialized returns (bool) {
+        // Delegator must be unbonded
+        require(delegatorStatus(msg.sender) == DelegatorStatus.Unbonded);
 
-            delete transcoders[msg.sender];
-        } else if (delegatorStatus(msg.sender) == DelegatorStatus.Unbonded){
-            token.transfer(msg.sender, delegators[msg.sender].bondedAmount);
+        token.transfer(msg.sender, delegators[msg.sender].bondedAmount);
 
-            delete delegators[msg.sender];
-        } else {
-            // Sender is neither a transcoder or delegator
-            revert();
-        }
+        delete delegators[msg.sender];
```

Comment:
Will this be a problem if someone withdraws, but then receives a reward or fee payout after the fact?

Reply:
If you are a delegator that unbonds and then withdraws, you shouldn't be eligible for any future rewards or fee payouts since you are no longer bonded to the transcoder. If you are a transcoder and you unbond/withdraw your stake, the next time you receive rewards or fee payouts, a delegator struct corresponding to your address will be populated with the correct amounts - essentially re-instantiating a bonded delegator